src/templates/paint_on_template_tool.cpp

```diff
@@ -75,16 +75,71 @@ void PaintOnTemplateTool::init()
 {
 	setStatusBarText(tr("<b>Click and drag</b>: Paint. <b>Right click and drag</b>: Erase. "));
 	
-	widget = new PaintOnTemplatePaletteWidget(false);
+	widget = makeToolBar();
 	editor->showPopupWidget(widget, tr("Color selection"));
-	connect(widget, &PaintOnTemplatePaletteWidget::colorSelected, this, &PaintOnTemplateTool::colorSelected);
-	connect(widget, &PaintOnTemplatePaletteWidget::undoSelected, this, &PaintOnTemplateTool::undoSelected);
-	connect(widget, &PaintOnTemplatePaletteWidget::redoSelected, this, &PaintOnTemplateTool::redoSelected);
-	colorSelected(widget->getSelectedColor());
 	
 	MapEditorTool::init();
 }
 
+ActionGridBar* PaintOnTemplateTool::makeToolBar()
+{
+	QSettings settings;
+	settings.beginGroup(QStringLiteral("PaintOnTemplateTool"));
+	auto last_selected = settings.value(QStringLiteral("selectedColor")).toString();
+	
+	auto* toolbar = new ActionGridBar(ActionGridBar::Horizontal, 2);
+	auto* color_button_group = new QButtonGroup(this);
+	
+	int count = 0;
+	static QColor const default_colors[8] = {
+	    qRgb(255,   0,   0),
+	    qRgb(255, 255,   0),
+	    qRgb(  0, 255,   0),
+	    qRgb(219,   0, 216),
+	    qRgb(  0,   0, 255),
+	    qRgb(209,  92,   0),
+	    qRgb(  0,   0,   0),
+	    qRgb(255, 255, 255)
+	};
+	for (auto const& color: default_colors)
+	{
+		QPixmap pixmap(1,1);
+		pixmap.fill(color);
+		auto* action = new QAction(pixmap, color.name(QColor::HexArgb), toolbar);
+		action->setCheckable(true);
+		toolbar->addAction(action, count % 2, count / 2);
+		color_button_group->addButton(toolbar->getButtonForAction(action));
+		if (count == 0 || action->text() == last_selected)
+		{
+			paint_color = color;
+			action->setChecked(true);
+		}
+		connect(action, &QAction::triggered, this, [this, color](bool checked) { if (checked) colorSelected(color); });
+		++count;
+	}
+	
+	auto* undo_action = new QAction(QIcon(QString::fromLatin1(":/images/undo.png")),
+	                                ::OpenOrienteering::MapEditorController::tr("Undo"),
+	                                toolbar);
+	connect(undo_action, &QAction::triggered, this, &PaintOnTemplateTool::undoSelected);
+	toolbar->addActionAtEnd(undo_action, 0, 0);
+	
+	auto* redo_action = new QAction(QIcon(QString::fromLatin1(":/images/redo.png")),
+	                                ::OpenOrienteering::MapEditorController::tr("Redo"),
+	                                toolbar);
+	connect(redo_action, &QAction::triggered, this, &PaintOnTemplateTool::redoSelected);
+	toolbar->addActionAtEnd(redo_action, 1, 0);
+	
+	auto* fill_action = new QAction(QIcon(QString::fromLatin1(":/images/scribble-fill-shapes.png")),
+	                                tr("Filled area"),
```

Comment:
This `tr()` lacks full qualification. I think it was required for `lupdate` or so. Might be worth fixing.

Reply:
I think it is okay. The full qualification is needed for other contexts than the the current class name IIRC. I will verify when this when updating translations.